model/src/lib.rs

```diff
@@ -20,6 +22,26 @@ pub struct Signature {
     pub v: u8,
 }
 
+pub trait EIP712Signing {
+    fn digest(&self) -> [u8; 32];
+    fn signature(&self) -> Signature;
+
+    fn sign_self_with(&self, domain_separator: &DomainSeparator, key: &SecretKeyRef) -> Signature {
+        let message = Signature::signing_digest_message(domain_separator, &self.digest());
+        // Unwrap because the only error is for invalid messages which we don't create.
+        let signature = Key::sign(key, &message, None).unwrap();
+        Signature {
+            v: signature.v as u8 | 0x80,
+            r: signature.r,
+            s: signature.s,
+        }
+    }
+
+    fn validate_signature(&self, domain_separator: &DomainSeparator) -> Option<H160> {
```

Comment:
Add comment here explaining that this returns the public key if the signature is well formed, to make it completely obvious.